### ISSUE TITLE
Pagination

### ISSUE THREAD

**Issue by ghost** [nifty-site-manager/nsm#19]
This is supported by other projects:

https://gohugo.io/templates/pagination

and it appears to not be supported currently:

> I haven't worked out a good way to do pagination yet

https://github.com/nifty-site-manager/nsm/issues/16#issuecomment-573378471

> One thing I haven't worked out yet is a good way for people to do pagination

https://github.com/nifty-site-manager/nsm/issues/11#issuecomment-572816705

**Comment by n-ham** [comment]
Yep pagination is a big one that needs fleshing out. I haven't worked out the details yet, but I'm guessing it will require being able to do something like

    @page(page-name)
    {
        page-text
    }

And having Nift do a separate page to the one that's being build with `page-text` added in at that spot. Eg. the page being built might be @[site-dir]/@[name]@[output-ext] and the extra page might be built as @[site-dir]/@[name]/@[page-name]@output-ext] or something like that? Would also need a good way to do eg. `@pathto(name)` but for paginated pages.

**Comment by cup** [comment]
Hm, I dont follow

when I say pagination, I mean this:

say its a blog, and you have 100 posts. Nift should be able to output for
example 10 pages listing 10 posts each, with just maybe Title, Date and Link to
each post.

**Comment by n-ham** [comment]
Yep that's the kind of thing I'm trying to flesh out above. Like you might do

```
@page(page/1)
{
    links for first page
}
```

So once there's `for` loops etc. you could work out how many pages you want and inside a `for` loop have `@page` calls. 

I'm not sure if that gives a good way to do links for each of the pages as well though? And not sure if it would even work nicely.. 

Am open to other suggestions as well.

**Comment by cup** [comment]
I cant offer much help, other than to show what Hugo does. Here is with no
pagination:

~~~
{{ range .Pages }}
   <a href="{{ .Permalink }}">{{ .Title }}</a>
{{ end }}
~~~

with pagination, default 10 per page:

~~~
{{ range (.Paginate .Pages).Pages }}
~~~

with pagination, 20 per page:

~~~
{{ range (.Paginate .Pages 20).Pages }}
~~~

with pagination, old to new:

~~~
{{ range (.Paginate .Pages.ByLastmod).Pages }}
~~~

with pagination, new to old:

~~~
{{ range (.Paginate .Pages.ByLastmod.Reverse).Pages }}
~~~

Then a template is inserted at the bottom, to generate the page navigation:

~~~
<footer>
   {{ template "_internal/pagination.html" . }}
</footer>
~~~

Source here:

https://github.com/gohugoio/hugo/blob/master/tpl/tplimpl/embedded/templates/pagination.html

**Comment by n-ham** [comment]
I will do something at some point, but it's unlikely going to be until I get more of the programming language features implemented sorry.

**Comment by n-ham** [comment]
Progress update: Sorry it's taken me so long to get on to this, I have been working hard on implementing those programming language features (there's now an in built template programming language and an in built scripting language that are very similar to each other). I have also embedded Lua (LuaJIT which is insanely fast and a fully fledged programming language), exprtk (a library for parsing and evaluating mathematical expressions) and chaiscript (slower than the other two but very nice for a few other reasons). I have also added a bunch of other things like colour to console output, progress of builds, etc. etc.. 

I am hopefully less than two weeks away from releasing all of this stuff, at which point I will then get started on a nice multithreaded pagination solution (though with the next release it'd be possible to roll your own solution too, but I am keen to implement a multithreaded solution that's inbuilt and super fast).
